contents/handbook/engineering/async-migrations.md

```diff
@@ -0,0 +1,222 @@
+            rollback="TRUNCATE TABLE table_x_new",
+        ),
+    ]
+    
+    # only run if the table doesn't already exist
+    def is_required(self):
+        result = sync_execute("SELECT count(*) FROM system.tables WHERE database='posthog' AND name='table_x_new'")
+        return result[0][0] == 0 
+
+    # check if we're not running low on storage
+    def healthcheck(self):
+        result = sync_execute("SELECT free_space FROM system.disks")
+        if result[0][0] < 1000000000:
+            return (False, "ClickHouse available storage below 1GB") 
+
+        return (True, None)
+
+    # update progress based on the total rows moved
+    def progress(self, _):
+        result = sync_execute(f"SELECT count(*) FROM table_x")
+        result2 = sync_execute(f"SELECT count(*) FROM table_x_new")
+        total_rows_to_move = result2[0][0]
+        total_rows_moved = result[0][0]
+
+        progress = 100 * total_rows_moved / total_rows_to_move
+        return progress
+
+```
+
+</details>
+
+
+### Workflow and architecture 
+
+#### Setup
+
+When the Django server boots up - a setup step for async migrations happens, which does the following:
+
+1. Imports all the migration definitions 
+2. Populates a dependencies map and in-memory record of migration definitions
+3. Creates a database record for each
+4. Check if all async migrations necessary for this PostHog version have completed (else don't start)
+5. Triggers a migration to run if `AUTO_START_SPECIAL_MIGRATIONS` is set and there's an uncompleted migration for this version
+
+#### Running a migration
+
+When a migration is triggered, the following happens:
+
+1. A task is dispatched to Celery to run this migration in the background
+2. The following basic checks are performed to ensure the migration can indeed run:
+    1. We're not over the concurrent migrations limit
+    2. The migration can be run with the current PostHog version
+    3. The migration is not already running
+    4. The service version requirements are met (e.g. X < ClickHouse version < Y)
+    5. The migration's `is_required` check passes
+    6. The migration's `healthcheck` passes
+    7. The migration's dependency (if any) has been completed
+3. We run through each of the operations in order
+4. Every 30 minutes, a Celery task performs a healthcheck, to ensure that:
+   1. The Celery process running the migration didn't crash
+   2. The migration's healthcheck still passes 
+
+> **Note:** Async migrations can also be run synchronously (i.e. not in Celery) using the async migrations CLI (WIP) or the Django shell.
+
+#### Stopping a migration
+
+A migration can be stopped by issuing a command via Celery's app control to terminate the process running the task. 
+
+#### Rollbacks
+
+If a migration is stopped for any reason (manual trigger or error), we will attempt to roll back the migration following the operations specified in reverse order from the last started operation. 
```

Comment:
how can I stop a migration without a rollback?

Reply:
you added a comment that I should add a flag for this and I will :D